### ISSUE TITLE
Add support for high level synthesis plugins to return a DAGCircuit

### ISSUE THREAD

**Issue by mtreinish** [Qiskit/qiskit#11681]
### What should we add?

Right now the `HighLevelSynthesisPlugin` interface only allows for a plugin to return a `QuantumCircuit`, however this needlessly forces a conversion when `HighLevelSynthesis` is run as when the pass is being used we're always in a dag context. We can't change the interface, but we should expand it to support both a dag and circuit return from a plugin and just normalize to a dagcircuit inside the pass. So for passes that use a dagcircuit we avoid the conversion.

**Comment by alexanderivrii** [comment]
I agree that this will be very useful. In addition, I would also like the plugins to return an `Operation` when possible (rather than a quantum circuit or a dag circuit with a single operation).
